File: fixtures/parser/statements.py
# case: pass
pass

# case: break
break

# case: continue
continue

# case: return_value
def f():
    return 1

# case: return_bare
def f():
    return

# case: while_loop
while a:
    b

# case: while_else
while a:
    b
else:
    c

# case: for_loop
for i in items:
    f(i)

# case: for_unpacking
for k, v in pairs:
    f(k)

# case: function_definition
def f(a, b):
    return a

# case: function_no_arguments
def f():
    pass

# case: function_default_argument
def f(a=1):
    pass

# case: function_annotated
def f(a: int) -> int:
    return a

# case: class_definition
class C:
    pass

# case: class_with_base
class C(Base):
    pass

# case: import
import os

# case: import_dotted
import os.path

# case: import_as
import numpy as np

# case: from_import
from os import path

# case: from_import_star
from os import *

# case: relative_import
from . import sibling

# case: del
del a

# case: global
global x

# case: nonlocal
nonlocal x

# case: assert
assert a

# case: assert_with_message
assert a, "boom"

# case: raise
raise ValueError

# case: raise_from
raise ValueError from err

# case: bare_raise
raise

# case: try_except
try:
    a
except E:
    b

# case: try_except_as
try:
    a
except E as err:
    b

# case: try_finally
try:
    a
finally:
    b

# case: try_except_else_finally
try:
    a
except E:
    b
else:
    c
finally:
    d

# case: with_statement
with open(f) as fh:
    pass

# case: with_multiple_items
with a as x, b as y:
    pass

# case: lambda
f = lambda x: x

# case: lambda_no_args
f = lambda: 1

# case: decorated_function
@dec
def f():
    pass

# case: async_function
async def f():
    await g()

# case: async_for
async def f():
    async for x in y:
        pass

# case: yield_value
def f():
    yield 1

# case: yield_from
def f():
    yield from g()

# case: match_statement
match x:
    case 1:
        pass

# case: semicolon_separated
a = 1; b = 2

# case: semicolon_trailing
a = 1;

# case: type_alias
type Alias = int

File: crates/pylentil_ast/tests/parsing/statements.rs
use super::*;
const F: &str = "parser/statements.py";

#[test]
fn parse_statements_pass() {
    assert_eq!(stmt(F, "pass"), PyStatement::Pass);
}

#[test]
fn parse_statements_break() {
    assert_eq!(stmt(F, "break"), PyStatement::Break);
}

#[test]
fn parse_statements_continue() {
    assert_eq!(stmt(F, "continue"), PyStatement::Continue);
}

#[test]
fn parse_statements_return_value() {
    assert_parses(F, "return_value");
}

#[test]
fn parse_statements_return_bare() {
    assert_parses(F, "return_bare");
}

#[test]
fn parse_statements_while_loop() {
    assert_eq!(
        stmt(F, "while_loop"),
        while_stmt(name("a"), vec![expr_stmt(name("b"))], vec![])
    );
}

#[test]
fn parse_statements_while_else() {
    assert_parses(F, "while_else");
}

#[test]
fn parse_statements_for_loop() {
    assert_eq!(
        stmt(F, "for_loop"),
        for_stmt(
            store(name("i")),
            name("items"),
            vec![expr_stmt(call(name("print"), vec![name("i")], vec![]))]
        )
    );
}

#[test]
fn parse_statements_for_unpacking() {
    assert_parses(F, "for_unpacking");
}

#[test]
fn parse_statements_function_definition() {
    assert_eq!(
        stmt(F, "function_definition"),
        function_def(
            "f",
            PyArguments {
                args: vec![arg("a"), arg("b")],
                ..Default::default()
            },
            vec![return_stmt(Some(name("a")))]
        )
    );
}

#[test]
fn parse_statements_function_no_arguments() {
    assert_eq!(
        stmt(F, "function_no_arguments"),
        function_def("f", PyArguments::default(), vec![PyStatement::Pass])
    );
}

#[test]
fn parse_statements_function_default_argument() {
    assert_parses(F, "function_default_argument");
}

#[test]
fn parse_statements_function_annotated() {
    assert_parses(F, "function_annotated");
}

#[test]
fn parse_statements_class_definition() {
    assert_eq!(
        stmt(F, "class_definition"),
        class_def("C", vec![PyStatement::Pass])
    );
}

#[test]
fn parse_statements_class_with_base() {
    assert_parses(F, "class_with_base");
}

#[test]
fn parse_statements_import() {
    assert_parses(F, "import");
}

#[test]
fn parse_statements_import_dotted() {
    assert_parses(F, "import_dotted");
}

#[test]
fn parse_statements_import_as() {
    assert_parses(F, "import_as");
}

#[test]
fn parse_statements_from_import() {
    assert_parses(F, "from_import");
}

#[test]
fn parse_statements_from_import_star() {
    assert_parses(F, "from_import_star");
}

#[test]
fn parse_statements_relative_import() {
    assert_parses(F, "relative_import");
}

#[test]
fn parse_statements_del() {
    assert_parses(F, "del");
}

#[test]
fn parse_statements_global() {
    assert_parses(F, "global");
}

#[test]
fn parse_statements_nonlocal() {
    assert_parses(F, "nonlocal");
}

#[test]
fn parse_statements_assert() {
    assert_parses(F, "assert");
}

#[test]
fn parse_statements_assert_with_message() {
    assert_parses(F, "assert_with_message");
}

#[test]
fn parse_statements_raise() {
    assert_parses(F, "raise");
}

#[test]
fn parse_statements_raise_from() {
    assert_parses(F, "raise_from");
}

#[test]
fn parse_statements_bare_raise() {
    assert_parses(F, "bare_raise");
}

#[test]
fn parse_statements_try_except() {
    assert_parses(F, "try_except");
}

#[test]
fn parse_statements_try_except_as() {
    assert_parses(F, "try_except_as");
}

#[test]
fn parse_statements_try_finally() {
    assert_parses(F, "try_finally");
}

#[test]
fn parse_statements_try_except_else_finally() {
    assert_parses(F, "try_except_else_finally");
}

#[test]
fn parse_statements_with_statement() {
    assert_parses(F, "with_statement");
}

#[test]
fn parse_statements_with_multiple_items() {
    assert_parses(F, "with_multiple_items");
}

#[test]
fn parse_statements_lambda() {
    assert_parses(F, "lambda");
}

#[test]
fn parse_statements_lambda_no_args() {
    assert_parses(F, "lambda_no_args");
}

#[test]
fn parse_statements_decorated_function() {
    assert_parses(F, "decorated_function");
}

#[test]
fn parse_statements_async_function() {
    assert_parses(F, "async_function");
}

#[test]
fn parse_statements_async_for() {
    assert_parses(F, "async_for");
}

#[test]
fn parse_statements_yield_value() {
    assert_parses(F, "yield_value");
}

#[test]
fn parse_statements_yield_from() {
    assert_parses(F, "yield_from");
}

#[test]
fn parse_statements_match_statement() {
    assert_parses(F, "match_statement");
}

#[test]
fn parse_statements_semicolon_separated() {
    assert_eq!(
        body(F, "semicolon_separated"),
        vec![
            assign(vec![store(name("a"))], int(1)),
            assign(vec![store(name("b"))], int(2))
        ]
    );
}

#[test]
fn parse_statements_semicolon_trailing() {
    assert_parses(F, "semicolon_trailing");
}

#[test]
fn parse_statements_type_alias() {
    assert_parses(F, "type_alias");
}
